network/network_interface.ml

```diff
@@ -25,299 +25,609 @@ let uri () = "file:" ^ !default_path
-	external get_mtu : debug_info -> name:iface -> int = ""
-	external get_capabilities : debug_info -> name:iface -> string list = ""
-	external is_connected : debug_info -> name:iface -> bool = ""
-	external is_physical : debug_info -> name:iface -> bool = ""
-	external has_vlan: debug_info -> name:iface -> vlan:int -> bool = ""
-	external bring_down : debug_info -> name:iface -> unit = ""
-	external set_persistent : debug_info -> name:iface -> value:bool -> unit = ""
-	external make_config : debug_info -> ?conservative:bool -> config:(iface * interface_config_t) list-> unit -> unit = ""
+  let replace_assoc key new_value existing =
+    (key, new_value) :: (List.filter (fun (k, _) -> k <> key) existing) in
+
+  if List.mem_assoc name config then begin
+    replace_assoc name data config
+  end else
+    (name, data) :: config
+
+type errors =
+  | Script_missing of string
+        [@doc [
+          "[Script_missing (script)] is reported ";
+          "if unable to find [script]"]]
+  | Script_error of (string * string) list
+        [@doc [
+          "[Script_error ([(key * value); ...])] is reported ";
+          "when error occurs when executing script, the [key] and [value] indicates ";
+          "the infomation about the script and the error"]]
+  | Read_error of string
+        [@doc [
+          "[Read_error (file)] is reported ";
+          "when error occurs when reading [file]"]]
+  | Write_error of string
+        [@doc [
+          "[Write_error (file)] is reported ";
+          "when error occurs when writing [file]"]]
+  | Not_implemented
+      [@doc [
+        "[Not_implemented] is reported ";
+        "if the interface is not implemented"]]
+  | Vlan_in_use of (string * int)
+        [@doc [
+          "[Vlan_in_use (bridge, vlan_id)] is reported ";
+          "when [vlan_id] on [bridge] is inuse"]]
+  | PVS_proxy_connection_error
+      [@doc [
+        "[PVS_proxy_connection_error] is reported ";
+        "when unable to connect PVS proxy"]]
+  | Unknown_error
+      [@doc [
+        "The default variant for forward compatibility."
+      ]]
+[@@default Unknown_error]
+[@@deriving rpcty]
+
+exception NetworkError of errors
+
+let err = Error.{
+    def = errors;
+    raiser = (function | e -> raise (NetworkError e));
+    matcher = function | NetworkError e -> Some e | _ -> None
+  }
+
+module API(R : RPC) = struct
+  open R
+
+  let description = Interface.{
+      name = "Network";
+      namespace = None;
+      description = [
+        "This interface is used by Xapi and networkd to manage ";
+        "Xenserver network bridges and devices .";
+      ];
+      version=(1,0,0);
+    }
+
+  let implementation = implement description
+
+  let debug_info_p = Param.mk ~description:[
+      "an uninterpreted string to associate with the operation."
+    ] Types.string
+
+  let unit_p = Param.mk Types.unit
+
+  (* external clear_state: unit -> unit = "" *)
```

Comment:
You don't need any of these comments, the type is explicit in the definitions and the old type can be recovered using git

Reply:
Yes, that used to help you review the code changes.
I will remove them now.

Reply:
Thanks :)